File: python/minimum_penalty_for_a_shop.py
class Solution:
    def bestClosingTime(self, customers: str) -> int:
        ps1: list[int] = [0] * (len(customers) + 1)
        ps2: list[int] = [0] * (len(customers) + 1)

        curr: int = 0
        for i in range(0, len(customers)):
            if customers[i] == "N":
                curr += 1
            ps1[i + 1] = curr

        curr = 0
        for i in reversed(range(0, len(customers))):
            if customers[i] == "Y":
                curr += 1
            ps2[i] = curr

        minP: int = float("inf")
        minIndex: int = 0
        for i in range(len(ps1)):
            if ps1[i] + ps2[i] < minP:
                minP = ps1[i] + ps2[i]
                minIndex = i

        return minIndex


def test_bestClosingTime_case_1():
    # arrange
    customers: str = "YYNY"
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.bestClosingTime(customers)

    # assert
    assert expected == actual


def test_bestClosingTime_case_2():
    # arrange
    customers: str = "NNNNN"
    expected: int = 0

    # act
    solution = Solution()
    actual = solution.bestClosingTime(customers)

    # assert
    assert expected == actual


def test_bestClosingTime_case_3():
    # arrange
    customers: str = "YYYY"
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.bestClosingTime(customers)

    # assert
    assert expected == actual

File: rust/src/code/minimum_penalty_for_a_shop.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn best_closing_time(customers: String) -> i32 {
        let customers: Vec<char> = customers.chars().collect();
        let mut ps1: Vec<i32> = vec![0; customers.len() + 1];
        let mut ps2: Vec<i32> = vec![0; customers.len() + 1];

        let mut penalty: i32 = 0;
        for i in 0..customers.len() {
            if customers[i] == 'N' {
                penalty += 1;
            }
            ps1[i + 1] = penalty;
        }

        penalty = 0;
        for i in (0..customers.len()).rev() {
            if customers[i] == 'Y' {
                penalty += 1;
            }
            ps2[i] = penalty;
        }

        let mut min_penalty: i32 = i32::MAX;
        let mut min_index: usize = 0;
        for i in 0..ps1.len() {
            if min_penalty > ps1[i] + ps2[i] {
                min_penalty = ps1[i] + ps2[i];
                min_index = i;
            }
        }

        return min_index as i32;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_best_closing_time_case_1() {
        // arrange
        let customers: String = "YYNY".to_string();
        let expected: i32 = 2;

        // act
        let actual = Solution::best_closing_time(customers);

        // assert
        assert_eq!(expected, actual);
    }
}
